PSR-6 cache support

boesing opened the pull request:
With the next major releases of `doctrine/dbal` and `doctrine/orm`, support for `doctrine/cache` will be dropped and only PSR-6 will be supported.

To prepare this repository for that change, I'd like to add support for PSR-6 caches in this PR.

### NOTE
This also drops support for various package versions:

- `doctrine/common` <3.2
- `doctrine/dbal` <3.3
- `doctrine/migrations` <3.4
- `doctrine/persistence` <2.3

boesing committed:
qa: add unit test do verify that the `CacheFactory` can return `CacheItemPoolInterface` instances

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
feature: allow PSR-6 annotation cache

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: always create `PsrCachedReader` for annotation readers

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: use `array_replace_recursive` for recursive array configuration merges

In `Roave\PsrContainerDoctrine\ConfigurationFactory`, the `doctrine/orm` configuration is assembled. That configuration has a nested array configuration for `second_level_cache` which also have defaults.
Due to the array concatenation, nested array values are not applied and thus the configured defaults within the `ConfigurationFactory` are not applied.

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
feature: use PSR-6 caches and methods for ORM configuration

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: drop some support for `doctrine/*` packages

- drops support for `doctrine/common` <3.2
- drops support for `doctrine/dbal` <3.3
- drops support for `doctrine/migrations` <3.4
- drops support for `doctrine/persistence` <2.3

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: remove conflict

doctrine already conflicts itself with versions of `doctrine/annotations` <1.13 in `doctrien/orm` 2.11.x

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing commented on psalm.xml.dist:
```diff
@@ -59,14 +54,13 @@
             </errorLevel>
         </DeprecatedClass>
-        <DeprecatedMethod>
+        <PropertyNotSetInConstructor>
```
Can be removed with #61 since `psalm/phpunit-plugin` treats `setUp` method the same way as `__construct`.

boesing commented on src/AbstractFactory.php:
```diff
@@ -83,7 +84,7 @@ protected function retrieveConfig(ContainerInterface $container, string $configK
 
         if (array_key_exists($configKey, $sectionConfig)) {
-            return $sectionConfig[$configKey] + $this->getDefaultConfig($configKey);
+            return array_replace_recursive($this->getDefaultConfig($configKey), $sectionConfig[$configKey]);
```
This is actually a bugfix which could also be passed to lower versions of this component.
I've written down some more details in the commit message:

> In `Roave\PsrContainerDoctrine\ConfigurationFactory`, the `doctrine/orm` configuration is assembled. That configuration has a nested array configuration for `second_level_cache` which also have defaults. 
> Due to the array concatenation, nested array values are not applied and thus the configured defaults within the `ConfigurationFactory` are not applied.

Ocramius replied:
Depending on how critical this is, we may backport it, but since nobody cried yet: :shrug: 

This will be released in a new minor anyway, so upgrading should be trivial.

boesing replied:
Yah, would say the same. If someone wants this in 3.3.x to make it work on older doctrine/* component versions (which we dropped in this patch) then this can be easily backported.

Slamdunk replied:
See https://github.com/Roave/psr-container-doctrine/pull/65

yethee replied:
This change breaks configuration for `doctrine/migrations`. It's not possible to drop the default `migrations_paths`.

Ocramius replied:
@yethee I just released `3.4.1` with @Slamdunk's fix.

Ocramius commented on composer.json:
```diff
@@ -18,18 +18,19 @@
         "doctrine/coding-standard": "^8.1",
         "phpunit/phpunit": "^9.4",
         "psalm/plugin-phpunit": "^0.9",
+        "psr/cache": "^1.0",
```
`psr/cache` is currently at `3.0` - we'd be moving to an older release here :thinking:

boesing replied:
Yah, also not sure if we actually need this dependency as it is a requirement by doctrine anyways. Will drop it here.

boesing commented on src/DriverFactory.php:
```diff
@@ -123,4 +129,32 @@ private function registerAnnotationLoader(): void
+        $cacheType = is_object($cache) ? get_class($cache) : gettype($cache);
+        assert($cacheType !== '');
+
+        throw InvalidArgumentException::fromUnsupportedCacheType(
```
This is actually new. Before this patch, we only got `TypeError` as the return type of `CacheFactory` is never validated.

Not sure why psalm is not yelling at it but it can return every type and we only see problems when it is passed to the `__construct` of the reader.

boesing commented on test/ConfigurationFactoryTest.php:
```diff
@@ -0,0 +1,82 @@
+                        'result_cache' => 'result',
+                        'query_cache' => 'query',
+                        'hydration_cache' => 'hydration',
+                        'second_level_cache' => ['enabled' => true],
```
This config actually led to the bugfix mentioned in https://github.com/Roave/psr-container-doctrine/pull/63/files#r806931084

boesing commented on test/ConfigurationFactoryTest.php:
```diff
@@ -0,0 +1,82 @@
+    private function exctractPropertyValue(object $object, string $propertyName)
+    {
+        $property = new ReflectionProperty($object, $propertyName);
+        if (PHP_VERSION_ID < 80000) {
```
Thanks to Marco, we do not need to set accessibility to reflection properties in PHP 8+

boesing replied:
whooot. Thats only for PHP 8.1+? 😱

Ocramius commented on src/CacheFactory.php:
```diff
@@ -26,7 +27,7 @@
 
 /**
- * @method Cache __invoke(ContainerInterface $container)
+ * @method Cache|CacheItemPoolInterface __invoke(ContainerInterface $container)
```
This seems like a BC Break - is this an `|` or an `&`?

boesing replied:
it is `|` but it depends on how you call this factory.
Since it is actually widening the return value it also requires the configuration of a project to change.
So it does not change anything for existing projects.

boesing committed:
bugfix: do not set accessibility for PHP 8.1+

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

Ocramius commented on src/ConfigurationFactory.php:
```diff
@@ -186,4 +214,22 @@ protected function getDefaultConfig(string $configKey): array
+            $cache = CacheAdapter::wrap($cache);
+        }
+
+        $psr($cache);
```
Rather than `$psr`, this is `$setCacheOnConfiguration`? :thinking:

Ocramius commented on src/DriverFactory.php:
```diff
@@ -123,4 +129,32 @@ private function registerAnnotationLoader(): void
+            return new PsrCachedReader($reader, $cache);
+        }
+
+        $cacheType = is_object($cache) ? get_class($cache) : gettype($cache);
```
I suggest moving this check into the exception named constructor to reduce the conditionals/complexity in here

Ocramius commented on src/Exception/InvalidArgumentException.php:
```diff
@@ -21,4 +24,19 @@ public static function forInvalidEventListenerConfig($listenerConfig): self
     }
+
+    /**
+     * @param non-empty-string $cacheType
```
`@param mixed $unsupportedCache` would clarify this

Ocramius commented on test/ConfigurationFactoryTest.php:
```diff
@@ -0,0 +1,82 @@
+    private function exctractPropertyValue(object $object, string $propertyName)
+    {
+        $property = new ReflectionProperty($object, $propertyName);
+        if (PHP_VERSION_ID < 80100) {
```
In theory, no need to add this conditional: the method shouldn't be deprecated yet, so calling it on PHP 8.1 is a no-op.

boesing committed:
qa: remove PHP 8.1+ property accessibility logic

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: reduce conditionals/complexity in drivery factory

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: rename method argument and change return value of method to `void`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: remove `psr/cache` requirement

This is also a requirement by `doctrine/*` components and thus can be omitted here since we do not implement that interface.

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

Ocramius reviewed (approved):
Thanks @boesing!
